### PR TITLE
feat: replace react router with tanstack router

### PR DESCRIPTION
**Pull request by MieszkoTSH**

### CODE REVIEW

File: src/api/mocks/handlers.ts [hunk 1]
```diff
@@ -48,7 +48,7 @@ const usersHandler = http.get<PathParams, DefaultBodyType, GetUsersResponse>('/u
     return HttpResponse.json({ users: allUsers }, { status: 200 });
   }
 
-  const start = page * count;
+  const start = (page - 1) * count;
```
Comment (tokerson) [inline src/api/mocks/handlers.ts:51]: why was this change needed after upgrading the router? It looks like the the router has an impact on the api

File: src/routeTree.gen.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+// @ts-nocheck
+
+// noinspection JSUnusedGlobalSymbols
+
+// This file is auto-generated by TanStack Router
+
+// Import Routes
+
+import { Route as rootRoute } from './routes/__root'
```
Comment (tokerson) [inline src/routeTree.gen.ts:13]: I understand that this file was autogenerated right? Could you explain a bit more based on what this was generated?

File: src/routes/users/-components/UsersList.tsx [hunk 3]
```diff
@@ -1,20 +1,17 @@
...
-
-  const sort = searchParams.get('sort') as SortType;
-  const page = searchParams.get('page');
+  const { sort, page } = useSearch({ from: '/users/' });
+  const navigate = useNavigate();
 
   const { data: usersResponse, isFetched: areUsersFetched } = useQuery(
     'getUsersList',
-    { page: page || undefined },
+    { page: page?.toString() || undefined },
```
Comment (tokerson) [inline src/routes/users/-components/UsersList.tsx:14]: I think it can be simplified
```suggestion
    { page: page?.toString() },
```

File: src/tests/index.tsx [hunk 4]
```diff
@@ -13,30 +20,24 @@ import { LocaleContext } from 'context/locale/localeContext/LocaleContext';
...
-  }
-
-  return <Router>{children}</Router>;
-};
-
 // @TODO: https://bitbucket.org/thesoftwarehouse/react-starter-boilerplate/pull-requests/5/rss-9-add-login-page/diff#comment-132626297
-const _Wrapper = ({ children, routerConfig = { withRouter: false } }: WrapperProps) => {
+const _Wrapper = ({ children, routerConfig }: WrapperProps) => {
   const [locale, setLocale] = useState<AppLocale>(defaultLocale);
+  const { routerPath = '/', currentPath = routerPath } = routerConfig || {};
```
Comment (tokerson) [inline src/tests/index.tsx:26]: My personal preference is to put a default value in props destructuring (on line 24) instead like it is done here using `|| {}` because thanks to that if you wanted to use routerConfig somewhere else in this component it would also be an empty object instead of `undefined`

File: src/ui/locationInfo/LocationInfo.tsx [hunk 5]
```diff
@@ -1,9 +1,9 @@
-import { useLocation } from 'react-router-dom';
+import { useRouterState } from '@tanstack/react-router';
 
 import { CodeBlock } from 'ui/codeBlock/CodeBlock';
 
 export const LocationInfo = () => {
-  const location = useLocation();
+  const location = useRouterState({ select: (state) => state.location });
```
Comment (tokerson) [inline src/ui/locationInfo/LocationInfo.tsx:6]: I think it may be useful to create a custom hook `useLocation` with that logic, cause I expect it may be helpful in more places in bigger apps.

File: src/routeTree.gen.ts (hunk 2 above)
Comment (MieszkoTSH) [reply src/routeTree.gen.ts:13]: This file is generated by `TanStackRouterVite` plugin. Basically it watches the files inside `routes` folder and whenever you create a route file like `index.tsx` it adds it to this file. https://tanstack.com/router/latest/docs/framework/react/guide/file-based-routing
https://tanstack.com/router/latest/docs/framework/react/decisions-on-dx#3-why-is-file-based-routing-the-preferred-way-to-define-routes

File: src/api/mocks/handlers.ts (hunk 1 above)
Comment (MieszkoTSH) [reply src/api/mocks/handlers.ts:51]: While changing the router I did set up default value for page to 1. Later on I discovered that before it was set to "0" but at least for me starting pagination with 1 makes more sense and might be more common(?)

Comment (tokerson) [reply src/api/mocks/handlers.ts:51]: I agree 👍

File: src/context/apiClient/apiClientContextController/interceptors/responseInterceptors.ts [hunk 6]
```diff
@@ -20,7 +19,7 @@ export const responseFailureInterceptor = async (error: AxiosError<unknown>) =>
     authStorage.expires = null;
     authStorage.refreshToken = null;
 
-    window.location.replace(AppRoute.login);
+    window.location.replace('/login');
```
Comment (MieszkoTSH) [inline src/context/apiClient/apiClientContextController/interceptors/responseInterceptors.ts:22]: I think this is not really optimal, as there is no typesaftey. But if we implement https://headstart.atlassian.net/browse/ZN-571 we could move it somewhere into react lifecycle as we would be subscribed to authStorage values

**Review by tokerson — APPROVED** [review]
